tests/defineEntity.test.ts

```diff
@@ -0,0 +1,1106 @@
+    const Address = defineEntity({
+      name: 'Address',
+      embeddable: true,
+      properties: p => ({
+        id: p.integer().primary().autoincrement(),
+        street: p.string(),
+        city: p.string(),
+        state: p.string(),
+        zip: p.string(),
+      }),
+    });
+
+    const Foo = defineEntity({
+      name: 'Foo',
+      properties: p => ({
+        id: p.integer().primary().autoincrement(),
+        name: p.string(),
+        address: p.embedded(Address),
+      }),
+    });
+
+    type IFoo = InferEntity<typeof Foo>;
+    type IAddress = InferEntity<typeof Address>;
+    assert<IsExact<IFoo, { id: number; name: string; address: IAddress }>>(true);
+
+    const AddressSchema = new EntitySchema({
+      name: 'Address',
+      embeddable: true,
+      properties: {
+        id: { type: types.integer, primary: true, autoincrement: true },
+        street: { type: types.string },
+        city: { type: types.string },
+        state: { type: types.string },
+        zip: { type: types.string },
+      },
+    });
+
+    const FooSchema = new EntitySchema({
+      name: 'Foo',
+      properties: {
+        id: { type: types.integer, primary: true, autoincrement: true },
+        name: { type: types.string },
+        address: { kind: 'embedded', entity: () => Address },
+      },
+    });
+
+    expect(Address.meta).toEqual(asSnapshot(AddressSchema.meta));
+    expect(Foo.meta).toEqual(asSnapshot(FooSchema.meta));
+  });
+
+  it('should define entity with many to one relation', () => {
+    const Foo = defineEntity({
+      name: 'Foo',
+      properties: p => ({
+        id: p.integer().primary().autoincrement(),
+        name: p.string(),
+        friend: () => p.manyToOne(Foo),
+      }),
+    });
+
+    type IFoo = InferEntity<typeof Foo>;
+    assert<IsExact<IFoo, { id: number; name: string; friend: Reference<IFoo> }>>(true);
+    assert<IsExact<UnwrapRef<UnwrapRef<UnwrapRef<IFoo['friend']>['friend']>['friend']>['name'], string>>(true);
+    assert<IsExact<UnwrapRef<UnwrapRef<UnwrapRef<IFoo['friend']>['friend']>['friend']>['name'], number>>(false);
+
+    const FooSchema = new EntitySchema<IFoo>({
+      name: 'Foo',
+      properties: {
+        id: { type: types.integer, primary: true, autoincrement: true },
+        name: { type: types.string },
+        friend: { kind: 'm:1', entity: () => Foo, ref: true },
+      },
+    });
+
+    expect(Foo.meta).toEqual(asSnapshot(FooSchema.meta));
+  });
+
+  it('should define entity with one to many relation', () => {
+    const Folder = defineEntity({
+      name: 'Folder',
+      properties: p => ({
+        id: p.integer().primary().autoincrement(),
+        name: p.string(),
+        files: () => p.oneToMany(File).mappedBy('folder'),
+      }),
+    });
+
+    const File = defineEntity({
+      name: 'File',
+      properties: p => ({
+        id: p.integer().primary().autoincrement(),
+        name: p.string(),
+        folder: () => p.manyToOne(Folder),
```

Comment:
this sounds more in line with how things work
```suggestion
        folder: p.manyToOne(() => Folder),
```

Reply:
Yes, `folder: p.manyToOne(() => Folder)` definitely looks better, but it will cause an error related to circular references in TypeScript, ultimately resulting in the relationship type not being correctly inferred. This issue also plagues [Valibot](https://valibot.dev/guides/other/) and [zod](https://zod.dev/?id=recursive-types).

`folder: () => p.manyToOne(Folder)` is a reluctant choice. Inspired by [zod v4](https://v4.zod.dev/v4#true-recursive-types), as long as we use the entire property as a callback function, we can bypass TypeScript's circular reference check.

Another approach is to define the property using a getter, but I think this syntax is not as concise as the callback function:
```ts
const File = defineEntity({
  name: 'File',
  properties: p => ({
    id: p.integer().primary().autoincrement(),
    name: p.string(),
    get folder() {
      return p.manyToOne(() => Folder),
    },
  }),
});
```